lib/backend/k8s/resources/ipam_affinity.go

```diff
@@ -0,0 +1,247 @@
+				APIVersion: "crd.projectcalico.org/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            name,
+				ResourceVersion: kvpv1.Revision,
+				Annotations: map[string]string{
+					"projectcalico.org/host": host,
+					"projectcalico.org/cidr": cidr,
+				},
+			},
+			Spec: apiv3.BlockAffinitySpec{
+				State: string(state),
+			},
+		},
+		Revision: kvpv1.Revision,
+	}
+}
+
+func (c *affinityBlockClient) Create(ctx context.Context, kvp *model.KVPair) (*model.KVPair, error) {
+	nkvp := c.toV3(kvp)
+	kvp, err := c.rc.Create(ctx, nkvp)
+	if err != nil {
+		return nil, err
+	}
+	v1kvp, err := c.toV1(kvp)
+	if err != nil {
+		return nil, err
+	}
+	return v1kvp, nil
+}
+
+func (c *affinityBlockClient) Update(ctx context.Context, kvp *model.KVPair) (*model.KVPair, error) {
+	nkvp := c.toV3(kvp)
+	kvp, err := c.rc.Update(ctx, nkvp)
+	if err != nil {
+		return nil, err
+	}
+	v1kvp, err := c.toV1(kvp)
+	if err != nil {
+		return nil, err
+	}
+	return v1kvp, nil
+}
+
+func (c *affinityBlockClient) Delete(ctx context.Context, key model.Key, revision string, uid *types.UID) (*model.KVPair, error) {
+	name, _, _ := c.v3Fields(key)
+	k := model.ResourceKey{
+		Name: name,
+		Kind: apiv3.KindBlockAffinity,
+	}
+	kvp, err := c.rc.Delete(ctx, k, revision, key.(model.BlockAffinityKey).UID)
+	if err != nil {
+		return nil, err
+	}
+	v1kvp, err := c.toV1(kvp)
+	if err != nil {
+		return nil, err
+	}
+	return v1kvp, nil
+}
+
+func (c *affinityBlockClient) Get(ctx context.Context, key model.Key, revision string) (*model.KVPair, error) {
+	name, _, _ := c.v3Fields(key)
+	k := model.ResourceKey{
+		Name: name,
+		Kind: apiv3.KindBlockAffinity,
+	}
+	kvp, err := c.rc.Get(ctx, k, revision)
+	if err != nil {
+		return nil, err
+	}
+	v1kvp, err := c.toV1(kvp)
+	if err != nil {
+		return nil, err
+	}
+	return v1kvp, nil
+}
+
+func (c *affinityBlockClient) List(ctx context.Context, list model.ListInterface, revision string) (*model.KVPairList, error) {
+	l := model.ResourceListOptions{Kind: apiv3.KindBlockAffinity}
+	v3list, err := c.rc.List(ctx, l, revision)
+	if err != nil {
+		return nil, err
+	}
+
+	host := list.(model.BlockAffinityListOptions).Host
+	requestedIPVersion := list.(model.BlockAffinityListOptions).IPVersion
+
+	kvpl := &model.KVPairList{KVPairs: []*model.KVPair{}}
+	for _, i := range v3list.KVPairs {
+		v1kvp, err := c.toV1(i)
+		if err != nil {
+			return nil, err
+		}
+		if host == "" || v1kvp.Key.(model.BlockAffinityKey).Host == host {
+			cidr := v1kvp.Key.(model.BlockAffinityKey).CIDR
+			cidr2 := &cidr
+			if requestedIPVersion == 0 || requestedIPVersion == cidr2.Version() {
+				// Matches the given host and IP version.
+				kvpl.KVPairs = append(kvpl.KVPairs, v1kvp)
+			}
+		}
+	}
+	return kvpl, nil
+}
+
+func (c *affinityBlockClient) Watch(ctx context.Context, list model.ListInterface, revision string) (api.WatchInterface, error) {
+	log.Warn("Operation Watch is not supported on AffinityBlock type")
```

Comment:
Don't we need to watch blocks in confd?

Reply:
I think we only watch block affinities in confd.

Reply:
woops, these are affinities!